quantum/color.c

```diff
@@ -113,9 +113,16 @@ RGB hsv_to_rgb_nocie(HSV hsv) {
 void convert_rgb_to_rgbw(rgb_led_t *led) {
     // Determine lowest value in all three colors, put that into
     // the white channel and then shift all colors by that amount
-    led->w = MIN(led->r, MIN(led->g, led->b));
-    led->r -= led->w;
-    led->g -= led->w;
-    led->b -= led->w;
+    if (led->r == led->g && led->r == led->b) {
+        led->w = led->r;
+        led->r = 0;
+        led->g = 0;
+        led->b = 0;
+    } else {
+        led->w = MIN(led->r, MIN(led->g, led->b));
+        led->r -= led->w;
+        led->g -= led->w;
+        led->b -= led->w;
+    }
```

Comment:
Does this actually change the behavior?  In the general case, if all RGB color components are equal, `MIN(led->r, MIN(led->g, led->b))` should give `led->r`, and then subtracting that value from the RGB components should give 0; so the special case should not be needed.

Reply:
That's why I've mentioned that this is a temporary workaround to the issue. Without `led->w = led->r;`, any LED's currently set to saturation `0` will not toggle off with RGB_TOG or sleep/timeout.

Reply:
This code doesn't really care about `RGB_TOG` or sleep, it simply converts the RGB value to RGBW. So perhaps it should actually be checking for the case where all three components are 0, rather than the same value?

Reply:
Update: I just tested this on a Proton C and an RGBW Neopixel stick - `UG_TOGG` turns the LEDs on and off just fine at any saturation level. I don't think there's anything wrong with the conversion code as it currently exists in the repo. The issue you're having must lie elsewhere, and this PR should be closed.

Reply:
I'll pull out some AVR controllers and my proton-c to continue testing I guess.

I'm only using SK6812's from BTF-Lighting right now so it might be a timing issue then..

I'll grab a RGBW neopixel to test too. Do you have any SK6812's you can test?

Thanks.

Reply:
As far as I'm aware, the [RGBW Neopixel stick](https://www.adafruit.com/product/2867) uses SK6812RGBWs, and I'm pretty sure they are the only 5050 WS2812-likes that have white LEDs.

Reply:
Thanks, I'll keep testing on my end. I also found this change improves the brightness of the white LED when saturation is set to `0` so there's that as well.

I realized I had the elite-c, not proton-c so all my testing would be on AVR if that would make any difference.

I'll close this PR if I can't reproduce the issue with any other controller+RGBW+rgb_matrix.

I previously tested it with a DZ60 (added rgb_matrix SK6812) and kbd67mkii/soldered (added rgb_matrix sk6812).

Reply:
Ah, I was testing with RGBLight (Underglow). I'll try RGB Matrix in the morning, it may be something to do with that.

Reply:
The real problem may be this piece of code in the `set_color` implementation for WS2812, which is from #21135 and tries to avoid a potentially expensive `ws2812_setleds()` call if there were no actual LED state changes:
https://github.com/qmk/qmk_firmware/blob/5acc3b2b7b13e6057623091a4bced7b1d9e76cdb/quantum/rgb_matrix/rgb_matrix_drivers.c#L180-L182
In the RGBW case `rgb_matrix_ws2812_array` contains RGBW data, and for the zero saturation the RGB components will be set to 0, so the above code will mistakenly ignore the request to set the RGB color to (0, 0, 0).

If we want to keep this optimization for RGBW LEDs, we would need to apply `convert_rgb_to_rgbw()` to a temporary buffer, and then compare all 4 components with `rgb_matrix_ws2812_array[i]` when detecting changes.

Reply:
Yep, I'm able to repro with RGB Matrix and it is probably due to the above.

Untangling this will be a little more complicated since the WS2812 driver is the only one with this single `set_leds()` API, rather than the `init`/`flush`/`set_color`/`set_color_all` setup, which necessitates this ugly glue code. Additionally, RGBLight will require some refactoring to remove its own internal LED array, as that should really be the the driver's responsibility. Split support also needs to be retained, ideally moved into the RGBLight & RGB Matrix code so that it "just works" for any driver.

All of this is probably a blocker for the original issue here to be resolved.